airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/cursor.py

```diff
@@ -0,0 +1,134 @@
+        pass
+
+    def close_partition(self, partition: Partition) -> None:
+        pass
+
+
+class ConcurrentCursor(Cursor):
+    # FIXME add unit tests
+    _START_BOUNDARY = 0
+    _END_BOUNDARY = 1
+
+    def __init__(
+        self,
+        stream_name: str,
+        stream_namespace: str,
+        stream_state: Any,
+        message_repository: MessageRepository,
+        connector_state_manager: ConnectorStateManager,
+        cursor_field: CursorField,
+        slice_boundary_fields: Optional[Tuple[str, str]],
+    ) -> None:
+        self._stream_name = stream_name
+        self._stream_namespace = stream_namespace
+        self._message_repository = message_repository
+        self._connector_state_manager = connector_state_manager
+        self._cursor_field = cursor_field
+        self._slice_boundary_fields = slice_boundary_fields if slice_boundary_fields else tuple()
+        self._most_recent_record: Optional[Record] = None
+
+        # TODO to migrate state. The migration should probably be outside of this class. Impact of not having this:
+        #  * Given a sync that emits no records, the emitted state message will be empty
+        self._state = {  # type: ignore  # waiting for when we implement the state migration
+            "slices": []  # empty for now but should look like `{start: 1, end: 10, parent_id: "id1"}`
+        }
+
+    def observe(self, record: Record) -> None:
+        if self._slice_boundary_fields:
+            # Given that slicing is done using the cursor field, we don't need to observe the record as we assume slices will describe what
+            # has been emitted. Assuming there is a chance that records might not be yet populated for the most recent slice, use a lookback
+            # window
+            return
+
+        if not self._most_recent_record or self._extract_cursor_value(self._most_recent_record) < self._extract_cursor_value(record):
+            self._most_recent_record = record
+
+    def _extract_cursor_value(self, record: Record) -> Comparable:
+        return self._cursor_field.extract_value(record)
+
+    def close_partition(self, partition: Partition) -> None:
+        slice_count_before = len(self._state["slices"])
+        self._add_slice_to_state(partition)
+        if slice_count_before < len(self._state["slices"]):
+            self._merge_partitions()
+            self._emit_state_message()
+
+    def _add_slice_to_state(self, partition: Partition) -> None:
+        partition_identifier = partition.identifier() or {}
+        if self._slice_boundary_fields:
+            self._state["slices"].append(
+                {
+                    "start": self._extract_from_slice(partition, self._slice_boundary_fields[self._START_BOUNDARY]),
+                    "end": self._extract_from_slice(partition, self._slice_boundary_fields[self._END_BOUNDARY]),
+                    **partition_identifier,
+                }
+            )
+        elif self._most_recent_record:
```

Comment:
This does not work at all as it assumes that all records up to the most recent records have been consumed but for Concurrent CDK, this is not the case. This seems to indicate that there is a concept of "finish sync" that is missing. This would have to include if there was an exception for any of the slices as if there is, we can't save the state